src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -114,21 +115,21 @@ public static Type fromId(byte id) throws ElasticsearchIllegalArgumentException
 
     private volatile Stage stage = Stage.INIT;
 
-    private Index index = new Index();
-    private Translog translog = new Translog();
-    private Start start = new Start();
-    private Timer timer = new Timer();
+    private final Index index = new Index();
+    private final Translog translog = new Translog();
+    private final Start start = new Start();
+    private final Timer timer = new Timer();
 
-    private Type type;
-    private ShardId shardId;
-    private RestoreSource restoreSource;
-    private DiscoveryNode sourceNode;
-    private DiscoveryNode targetNode;
+    private volatile Type type;
```

Comment:
I wonder if we can put all these in a struct like class where everything is final and make the struct volatile? the all seems to be set at the same time?